### PR TITLE
New version

### PR DESCRIPTION
**Pull request by zkamvar**
Here is the proposed new version of epiflows. Thanks to everyone for being patient. This implements a class that's based off of the [epicontacts](http://www.repidemicsconsortium.org/epicontacts/) class with the rationale that flows can be thought of as contacts between locations. I've added introductory content to the README, the introduction vignette, and created an epiflows-class vignette for demonstration of how to create an epiflows object from two data frames.

Please try it out and let me know what you think:

```r
devtools::install_github("reconhub/epiflows@new_version", build_vignettes = TRUE)
```

My only qualm with the way I have it right now is that the subsetting is not terribly intuitive because it's based on epicontacts where the linelist is the primary consideration. I had attempted to override the behavior, but that broke compatibility with epicontacts functions like `thin()`, causing unexpected (sometimes silent) errors. Because of that, I was comfortable having a less-than-intuitive subsettor. 

Below are more detailed changes

------------------------

## Changes to class

* epiflows class inherits the epicontacts class. This class is built from two
  data frames describing a linlist (referred here as `locations`) and contacts
  (referred here as `flows`). This was chosen over the previous implementation
  of two named vectors of inward and outward flows with a data frame of 
  locations to allow the user to flexibly store more than one focus. An extra
  element called `vars` has been added, storing a dictionary of variables that
  are present in the data frame for use with `estimate_risk_spread()`. These
  variables are accessible via `global_vars()`
  
## Changes to estimate_risk_spread

* The function is now a generic with a default and an epiflows method.
* The number of options have been reduced by combining the function and parameter
  arguments into one.
* All errors are collected and reported if multiple arguments are missing or misspelled
* Default varaibles have been set
* a new parameter, `return_all_simulations` has been added.

## Changes to constructor

* `make_epiflows()` is now a generic with methods for data frame, integer, and 
  numeric input.
  
## Changes to plotting

* `plot()` defaults to an interactive network plot from `visNetwork()` if no
  coordinates are available.
* `map_epiflows()` places flows on a map
* `vis_epiflows()` places flows on a network
* `grid_epiflows()` places flows on a bubble plot/grid

## New functions

* `get_flows()` returns the flows data frame optionally specifying flows from/to a given location
* `get_locations()` returns the locations data frame
* `get_coordinates()` returns coordinates or `NULL` from the locations data frame
* `get_id()` returns the identifier of all locations
* `get_n()` returns a numeric vector of cases from/to a given location
* `get_pop_size()` returns a vector of population sizes for locations
* `get_vars()` returns a specified variable from the locations data frame
* `set_vars()` allows you to set global variables to columns in the locations data frame
* `global_vars()` can return, set, and reset globally recognized variables
* `as.SpatialLinesDataFrame()` converts an epiflows object with coordinates to
  a `SpatialLinesDataFrame` class from the *sp* package.
  
## New data sets

* `Brazil_epiflows` is an epiflows object created from the `YF_Brazil` data
* `YF_flows` is the data frame of flows from `YF_Brazil`
* `YF_locations` is the data frame of locations from `YF_Brazil`
* `YF_coordinates` are the coordinates for the locations in `YF_Brazil`

## Removed functions

* `get_codes()` has been removed in favor of `get_id()`
* `get_flow_data()` has been removed in favor of `get_n()` and `get_flows()`
* `get_location_data()` has been removed in favor of `get_locations()`

## Misc

* `add_coordinates()` can now take a data frame input
* Magrittr pipes are no longer imported. 
* visNetwork is imported
* vdiffr is used for visual tests
* continuous integration and automated tests have been set up
* A new vignette describing the epiflows class has been added
* README and introduction vignette have been updated.

### CODE REVIEW

**Comment by zkamvar** [comment]
Oh, and coverage is currently at 93%: https://codecov.io/gh/reconhub/epiflows/branch/new_version

**Comment by Paula-Moraga** [comment]
Should length_of_stay be a separate vector (not included as a column of YF_locations) so different vectors length_of_stay are allowed?

I am asking this because in the Brazil data length_of_stay of residents of foreign countries in Brazil is the same for all states of Brazil, and length_of_stay is a column of YF_locations. But in other examples it is possible that length_of_stay is different for different states. For example, now length_of_stay of Argentina is 10.9 in Espirito Santo, Minas Gerais, etc. But in other examples we can have that length_of_stay is 10 for Espirto Santo, 8 for Minas Gerais, etc.

**Comment by zkamvar** [comment]
> Should length_of_stay be a separate vector (not included as a column of YF_locations) so different vectors length_of_stay are allowed?

I think this is a good point and brings up a bit of the difficulty with trying to create an all-encompassing class for these data. That being said, I think a solution exists with the current implementation. Because the locations data frame can contain any number of columns, the user can have a separate column determining the length of stay for each origin and simply switch which column the global variable `duration_stay` is pointing to, either in the object or in the function call. I'm imagining something like this:

<details>
<summary>Loading and setting up data</summary>

``` r
library("epiflows")
#> epiflows is loaded with the following global variables in `global_vars()`:
#> coordinates, pop_size, duration_stay, first_date, last_date, num_cases
data(YF_flows)
data(YF_locations)
data(YF_Brazil)
los <- matrix(YF_Brazil$length_of_stay, 
              nrow = length(YF_Brazil$length_of_stay),
              ncol = 5,
              dimnames = list(
                names(YF_Brazil$length_of_stay),
                paste("los", 
                      abbreviate(rownames(YF_Brazil$T_D), minlength = 2), 
                      sep = "_")
              ))
set.seed(500)
los[] <- replicate(5, sort(round(rnbinom(10, 4, mu = 16) + runif(nrow(los)), 1), decreasing = TRUE))
los
#>                          los_ES los_MG los_RdJ los_SP los_SB
#> Italy                      45.8   33.6    31.1   24.7   36.5
#> Spain                      31.1   31.0    28.6   20.9   31.1
#> Portugal                   27.0   30.1    19.3   15.9   26.0
#> Germany                    26.4   24.5    15.7   15.8   24.5
#> United Kingdom             24.8   22.0    14.9   14.6   18.3
#> United States of America   21.8   17.1    13.2   14.0   15.0
#> Argentina                  20.0   14.3    13.0   13.1   12.5
#> Chile                      13.0    9.4    11.9    7.2   12.1
#> Uruguay                    12.9    8.7     9.8    6.8    7.3
#> Paraguay                    8.7    3.0     8.8    5.3    3.9
losdf <- as.data.frame(los)
losdf$location_code <- rownames(losdf)
(locations <- merge(YF_locations, losdf, by = "location_code", all = TRUE))
#>               location_code location_population num_cases_time_window
#> 1            Espirito Santo             3973697                  2600
#> 2              Minas Gerais            20997560                  4870
#> 3            Rio de Janeiro            16635996                   170
#> 4                 Sao Paulo            44749699                   200
#> 5          Southeast Brazil            86356952                  7840
#> 6                 Argentina                  NA                    NA
#> 7                     Chile                  NA                    NA
#> 8                   Germany                  NA                    NA
#> 9                     Italy                  NA                    NA
#> 10                 Paraguay                  NA                    NA
#> 11                 Portugal                  NA                    NA
#> 12                    Spain                  NA                    NA
#> 13           United Kingdom                  NA                    NA
#> 14 United States of America                  NA                    NA
#> 15                  Uruguay                  NA                    NA
#>    first_date_cases last_date_cases length_of_stay los_ES los_MG los_RdJ
#> 1        2017-01-04      2017-04-30             NA     NA     NA      NA
#> 2        2016-12-19      2017-04-20             NA     NA     NA      NA
#> 3        2017-02-19      2017-05-10             NA     NA     NA      NA
#> 4        2016-12-17      2017-04-20             NA     NA     NA      NA
#> 5        2016-12-17      2017-05-10             NA     NA     NA      NA
#> 6              <NA>            <NA>           10.9   20.0   14.3    13.0
#> 7              <NA>            <NA>           10.3   13.0    9.4    11.9
#> 8              <NA>            <NA>           22.3   26.4   24.5    15.7
#> 9              <NA>            <NA>           30.1   45.8   33.6    31.1
#> 10             <NA>            <NA>            7.3    8.7    3.0     8.8
#> 11             <NA>            <NA>           27.2   27.0   30.1    19.3
#> 12             <NA>            <NA>           27.2   31.1   31.0    28.6
#> 13             <NA>            <NA>           19.5   24.8   22.0    14.9
#> 14             <NA>            <NA>           18.5   21.8   17.1    13.2
#> 15             <NA>            <NA>            8.0   12.9    8.7     9.8
#>    los_SP los_SB
#> 1      NA     NA
#> 2      NA     NA
#> 3      NA     NA
#> 4      NA     NA
#> 5      NA     NA
#> 6    13.1   12.5
#> 7     7.2   12.1
#> 8    15.8   24.5
#> 9    24.7   36.5
#> 10    5.3    3.9
#> 11   15.9   26.0
#> 12   20.9   31.1
#> 13   14.6   18.3
#> 14   14.0   15.0
#> 15    6.8    7.3
## Use both to create the epiflows object.
ef <- make_epiflows(flows         = YF_flows, 
                    locations     = locations, 
                    pop_size      = "location_population",
                    duration_stay = "length_of_stay",
                    num_cases     = "num_cases_time_window",
                    first_date    = "first_date_cases",
                    last_date     = "last_date_cases"
)
```

</details>


```r
set.seed(5000)
res <- estimate_risk_spread(ef, 
                            location_code          = "Espirito Santo",
                            r_incubation           = function(n) rlnorm(n, 1.46, 0.35),
                            r_infectious           = function(n) rnorm(n, 4.5, 1.5/1.96),
                            n_sim                  = 1e5,
                            return_all_simulations = TRUE)
#> Exportations done
#> Importations done
boxplot(res, horizontal = TRUE, las = 2)
```



``` r
get_vars(ef, "duration_stay", vector = TRUE)
#>           Espirito Santo             Minas Gerais           Rio de Janeiro 
#>                       NA                       NA                       NA 
#>                Sao Paulo         Southeast Brazil                Argentina 
#>                       NA                       NA                     10.9 
#>                    Chile                  Germany                    Italy 
#>                     10.3                     22.3                     30.1 
#>                 Paraguay                 Portugal                    Spain 
#>                      7.3                     27.2                     27.2 
#>           United Kingdom United States of America                  Uruguay 
#>                     19.5                     18.5                      8.0
```

Let's say we have a column in the data frame called `los_RdJ` to indicate the
length of stay in other countries from Rio de Janeiro.

```r
get_vars(ef, "los_RdJ", vector = TRUE)
#>           Espirito Santo             Minas Gerais           Rio de Janeiro 
#>                       NA                       NA                       NA 
#>                Sao Paulo         Southeast Brazil                Argentina 
#>                       NA                       NA                     13.0 
#>                    Chile                  Germany                    Italy 
#>                     11.9                     15.7                     31.1 
#>                 Paraguay                 Portugal                    Spain 
#>                      8.8                     19.3                     28.6 
#>           United Kingdom United States of America                  Uruguay 
#>                     14.9                     13.2                      9.8
```

Currently it can be reset by manipulating the `$vars$duration_stay` varaible
directly (though, I may want to make an accessor to make it safe).

```r
ef$vars$duration_stay <- "los_RdJ" # setting length of stay to Rio de Janeiro
get_vars(ef, "duration_stay", vector = TRUE) # Now the values have changed
#>           Espirito Santo             Minas Gerais           Rio de Janeiro 
#>                       NA                       NA                       NA 
#>                Sao Paulo         Southeast Brazil                Argentina 
#>                       NA                       NA                     13.0 
#>                    Chile                  Germany                    Italy 
#>                     11.9                     15.7                     31.1 
#>                 Paraguay                 Portugal                    Spain 
#>                      8.8                     19.3                     28.6 
#>           United Kingdom United States of America                  Uruguay 
#>                     14.9                     13.2                      9.8
set.seed(5000)
res <- estimate_risk_spread(ef, 
                            location_code          = "Rio de Janeiro", # set to Rio de Janeiro
                            r_incubation           = function(n) rlnorm(n, 1.46, 0.35),
                            r_infectious           = function(n) rnorm(n, 4.5, 1.5/1.96),
                            n_sim                  = 1e5,
                            return_all_simulations = TRUE)
#> Exportations done
#> Importations done
boxplot(res, horizontal = TRUE, las = 2)
```



Created on 2018-07-27 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by zkamvar** [comment]
moreover, it is still possible to override the object's internal variables by specifying them as arguments to `estimate_risk_spread()`:

```r
(short_holidays <- rnbinom(10, 5, mu = 5) + round(runif(10), 1))
#>  [1]  6.4  6.4  4.2  5.5  2.5  1.5  6.8  3.0 19.5  0.8
set.seed(5000)
res <- estimate_risk_spread(ef, 
                            location_code          = "Rio de Janeiro",
                            r_incubation           = function(n) rlnorm(n, 1.46, 0.35),
                            r_infectious           = function(n) rnorm(n, 4.5, 1.5/1.96),
                            n_sim                  = 1e5,
                            avg_length_stay_days   = short_holidays, # change the length of stay here
                            return_all_simulations = TRUE)
#> Exportations done
#> Importations done
boxplot(res, horizontal = TRUE, las = 2)
```

**Comment by Paula-Moraga** [comment]
I find complicated to show the user how to construct YF_locations with many NAs in the middle. I prefer to have length_of_stay as a separate vector as in the example that you use the vector short_holidays. What do you think?

**Comment by zkamvar** [comment]
> I find complicated to show the user how to construct YF_locations with many NAs in the middle. I prefer to have length_of_stay as a separate vector as in the example that you use the vector short_holidays. What do you think?

Either way works for me. In the example I showed, there was a hidden data setup section that assumed the user had a separate data frame/matrix with the different length_of_stays and used `merge()` to combine the locations and that data frame together:

```r
los
#>                          los_ES los_MG los_RdJ los_SP los_SB
#> Italy                      45.8   33.6    31.1   24.7   36.5
#> Spain                      31.1   31.0    28.6   20.9   31.1
#> Portugal                   27.0   30.1    19.3   15.9   26.0
#> Germany                    26.4   24.5    15.7   15.8   24.5
#> United Kingdom             24.8   22.0    14.9   14.6   18.3
#> United States of America   21.8   17.1    13.2   14.0   15.0
#> Argentina                  20.0   14.3    13.0   13.1   12.5
#> Chile                      13.0    9.4    11.9    7.2   12.1
#> Uruguay                    12.9    8.7     9.8    6.8    7.3
#> Paraguay                    8.7    3.0     8.8    5.3    3.9
losdf <- as.data.frame(los)
losdf$location_code <- rownames(losdf)
locations <- merge(YF_locations, losdf, by = "location_code", all = TRUE)
```

This was effectively what was done to convert the `YF_Brazil` data to `Brazil_epiflows`

```r
data("YF_Brazil")

# Create the flows data frame
from  <- as.data.frame.table(YF_Brazil$T_D, stringsAsFactors = FALSE)
to    <- as.data.frame.table(t(YF_Brazil$T_O), stringsAsFactors = FALSE)
flows <- rbind(from, to)
colnames(flows) <- c("from", "to", "n")

## Create the locations data frame
locations <- YF_Brazil$states
los       <- data.frame(location_code = names(YF_Brazil$length_of_stay), 
                        length_of_stay = YF_Brazil$length_of_stay)
locations <- merge(locations, los, by = "location_code", all = TRUE)

```

**Comment by Paula-Moraga** [comment]
Should we specify length_of_stay as a separate vector then? Will the plotting functions work as well?

**Comment by zkamvar** [comment]
> Should we specify length_of_stay as a separate vector then?

In the object? No. I think anything that is metadata related to a given location belongs in the locations data frame and anything that is metadata related to flows belongs in the flows data frame. It keeps things in one place and thus avoids errors from multiple objects swarming around. 

> Will the plotting functions work as well?

The plotting functions largely depend on the flows part of the object (except for the map, which needs coordinates), so they should work regardless.

**Comment by Paula-Moraga** [comment]
OK, so we leave everything as it is now?

We tell the user to construct two data.frames `locations` and `los`,
execute `locations <- merge(locations, los, by = "location_code", all = TRUE)`,
and then call `make_epiflows()` passing `locations`?

**Comment by zkamvar** [comment]
> OK, so we leave everything as it is now?

Yes and no. I think we can leave the class defined as it is, but add some more documentation and accessors to allow for easier and more intuitive switching between variables. 

> We tell the user to construct two data.frames locations and los,
execute locations <- merge(locations, los, by = "location_code", all = TRUE),
and then call make_epiflows() passing locations?

In this use case, yes (along with the flows data frame).

**Comment by zkamvar** [comment]
To be clear, `estimate_risk_spread()` still works without an epiflows object:

``` r
library("epiflows")
#> epiflows is loaded with the following global variables in `global_vars()`:
#> coordinates, pop_size, duration_stay, first_date, last_date, num_cases
data(YF_Brazil)
indstate <- 1 # "Espirito Santo" (indstate = 1), 
# "Minas Gerais" (indstate = 2), 
# "Southeast Brazil" (indstate = 5)

## functions generating incubation and infectious periods
incubation <- function(n) {
  rlnorm(n, 1.46, 0.35)
}

infectious <- function(n) {
  rnorm(n, 4.5, 1.5/1.96)
}

res <- estimate_risk_spread(
  location_code = YF_Brazil$states$location_code[indstate], 
  location_population = YF_Brazil$states$location_population[indstate], 
  num_cases_time_window = YF_Brazil$states$num_cases_time_window[indstate], 
  first_date_cases = YF_Brazil$states$first_date_cases[indstate], 
  last_date_cases = YF_Brazil$states$last_date_cases[indstate],
  num_travellers_to_other_locations = YF_Brazil$T_D[indstate,],
  num_travellers_from_other_locations = YF_Brazil$T_O[indstate,],
  avg_length_stay_days = YF_Brazil$length_of_stay,
  r_incubation = incubation,
  r_infectious = infectious,
  n_sim = 100000,
  return_all_simulations = FALSE
)
#> Exportations done
#> Importations done
head(res)
#>                          mean_cases lower_limit_95CI upper_limit_95CI
#> Italy                     0.2250291        0.1523650        0.3296226
#> Spain                     0.2274682        0.1540165        0.3331953
#> Portugal                  0.2316221        0.1568290        0.3392799
#> Germany                   0.1863519        0.1261771        0.2729683
#> United Kingdom            0.1768380        0.1197369        0.2590359
#> United States of America  0.9250067        0.6263238        1.3549733
```

Created on 2018-07-27 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by Paula-Moraga** [comment]
OK. If it is well documented there will be no problems. I am going to review the plotting functions and let you know my doubts. Many thanks!

**Comment by zkamvar** [comment]
I do have one question regarding behavior:

Would it be more beneficial or distracting for `estimate_risk_spread()` to print out the parameters it used?

**Comment by Paula-Moraga** [comment]
I think showing infectious and incubation distributions is OK. But showing cases, length_of_stay, population flows, etc is too much data. It will be distracting.

**Comment by Paula-Moraga** [comment]
I cannot plot the leaflet map because when I do `add_coordinates(ef)` I get this:
Warning messages:
1: geocode failed with status OVER_QUERY_LIMIT, location = "Sao Paulo" 
2: geocode failed with status OVER_QUERY_LIMIT, location = "Germany" 
3: geocode failed with status OVER_QUERY_LIMIT, location = "Portugal" 
4: geocode failed with status OVER_QUERY_LIMIT, location = "Spain" 
5: geocode failed with status OVER_QUERY_LIMIT, location = "Uruguay" 

Can we put long/lat in the YF_Brazil data so everybody can plot the leaflet map without need to do `add_coordinates(ef)`?

**Comment by zkamvar** [comment]
I've already done that as the data set `YF_coordinates`. I'll update the documentation with that instead of the MX one. 

```r
library("epiflows")
data("Brazil_epiflows")
data("YF_coordinates")
ef <- add_coordinates(Brazil_epiflows, YF_coordinates[-1])
plot(ef)
```

**Comment by zkamvar** [comment]
The example is also in the introduction vignette

**Comment by Paula-Moraga** [comment]
Nice!

**Comment by zkamvar** [comment]
I've added a new function called `set_vars()`. This allows you to set the global variables to different columns in your data frame should the need arise.

```r
get_vars(ef)$duration_stay
## [1] "length_of_stay_ES"
set_vars(ef, "duration_stay") <- "length_of_stay_MG"
get_vars(ef)$duration_stay
## [1] "length_of_stay_MG" 
```

**Comment by Paula-Moraga** [comment]
Very good work. Should I start updating the paper based on what you put in the vignettes?

**Comment by zkamvar** [comment]
Thank you ^_^

I think it may be safe to do so. I'll also make some changes to the paper after the weekend. 

I believe everything in the PR needs to be approved by @cosi1 as he is the maintainer.

File: R/add_coordinates.R
Comment (cosi1) [inline R/add_coordinates.R]: I'm not sure what our guidelines say about it, but I, personally, prefer putting one argument per line for longer argument lists, especially when they default to certain values.

That's a matter of preference, of course :-)

File: R/get_pop_size.R
Comment (cosi1) [inline R/get_pop_size.R]: `return()` can be omitted here.

File: R/map_epiflows.R
Comment (cosi1) [inline R/map_epiflows.R]: Assignments to `center` could be moved before the initial `if()` clause (`center <- if (is.character...)`) to highlight the purpose of this block. (`stop()` wouldn't matter here since the function would throw an error at this point anyway.)

File: R/plot.epiflows.R
Comment (cosi1) [inline R/plot.epiflows.R]: How about a `switch()` block here (without `return()`s)?

**Review by cosi1 — APPROVED** [review]

**Comment by cosi1** [comment]
Hello @zkamvar, all,

apologies for the delayed response, but I've just returned home after two weeks abroad.

Reviewed + accepted. Good job, guys!

-p-

File: R/add_coordinates.R
Comment (zkamvar) [reply R/add_coordinates.R]: I agree with that style choice and I'll make that change here. Thanks! :+1:

File: R/map_epiflows.R
Comment (zkamvar) [reply R/map_epiflows.R]: One of the problems here is that setView will render a blank map if the `lon` and `lat` arguments are `NA` or `NULL`, but not throw any perceivable error.

That being said, I can see how I can change it to avoid the nesting and meaningless assignment:

```r
  if (is.character(center) && length(center) == 1) {
    center <- get_coordinates(x, center)
  }
  if (!is.numeric(center) || length(center) != 2) {
    stop("center must be a single character string to use for ID lookup or a set of coordinates")
  }
  if (!is.null(center) && !all(is.finite(center))) {
    graph <- leaflet::setView(graph,
                              lng = center[[1]],
                              lat = center[[2]],
                              zoom = 2
                             )
  }
```

File: R/plot.epiflows.R
Comment (zkamvar) [reply R/plot.epiflows.R]: Agreed!

**Comment by zkamvar** [comment]
@Paula-Moraga and @cosi1, thank you both for the review and helpful comments. I'll go ahead and merge this as @cos1 has given it the green light.

**Comment by Paula-Moraga** [comment]
We need to remove Mex_travel_2009.RData from the package. We used it at the beginning but we do not have permission to publish that data.

**Comment by zkamvar** [comment]
That's a very good point. There are still a couple of examples that use it, but I can rework them with the Brazil data.

**Pull request merged**
